CaloCluster/src/CaloClusterFast_module.cc

```diff
@@ -1,322 +1,197 @@
+  void CaloClusterFast::makeClusters(CaloClusterCollection& caloClusters, const art::Handle<CaloHitCollection>& caloHitsHandle)
+  {
+      const Calorimeter& cal = *(GeomHandle<Calorimeter>());
+      const CaloHitCollection& caloHits(*caloHitsHandle);
+      if (caloHits.empty()) return;
+
+      std::vector<int> hits(caloHits.size());
+      std::iota(hits.begin(),hits.end(),0);
+      
+      auto functorTime = [&caloHits](int a, int b) {return caloHits[a].time() < caloHits[b].time();};
+      std::sort(hits.begin(),hits.end(),functorTime);
+      for (auto val : hits) if (caloHits[val].energyDep() < EnoiseCut_) val = -1;
+
+      auto iterSeed = hits.begin();
+      while (iterSeed != hits.end())
+      {
+          //find the first hit above the energy threshold, and the last hit within the required time window
+          const CaloHit& hitSeed = caloHits[*iterSeed];
+          if (*iterSeed==-1 || hitSeed.energyDep()< EminSeed_) {++iterSeed; continue;}
+          double timeStart = hitSeed.time();
+
+          //find the range around the seed time to search for other hits to form clusters
+          auto iterStart(iterSeed), iterStop(iterSeed);
+          while (iterStop  != hits.end()   && (*iterStop==-1  || caloHits[*iterStop].time() - timeStart < deltaTime_))  ++iterStop;
+          while (iterStart != hits.begin() && (*iterStart==-1 || timeStart - caloHits[*iterStart].time() < deltaTime_)) --iterStart;
+          ++iterStart; 
+
+          //start the clustering algorithm for the hits between iStart and iStop
+          std::queue<int> crystalToVisit;
+          std::vector<bool> isVisited(cal.nCrystal()); 
+
+          //put the first hit in the cluster list
+          std::vector<int> clusterList{*iterSeed};
+          int seedId = caloHits[*iterSeed].crystalID();
+          crystalToVisit.push(seedId);
+          *iterSeed=-1;
+
+          // loop until all seeds are processed
+          while (!crystalToVisit.empty())
+          {            
+              int visitId = crystalToVisit.front();
+              isVisited[visitId]=true;
+
+              std::vector<int>  neighborsId = cal.crystal(visitId).neighbors();
+              if (extendSearch_) std::copy(cal.nextNeighbors(visitId).begin(), cal.nextNeighbors(visitId).end(), std::back_inserter(neighborsId));
+
+              for (auto& iId : neighborsId)
+              {
+                  if (isVisited[iId]) continue;
+                  isVisited[iId] = true;
+
+                  //loop over the caloHits, check if one is in the neighbor list and add it to the cluster
+                  for (auto it=iterStart; it != iterStop; ++it)
+                  {
+                      if (*it==-1) continue;
+                      if (caloHits[*it].crystalID() != iId) continue;
+
+                      if (caloHits[*it].energyDep() > ExpandCut_) crystalToVisit.push(iId);
+                      clusterList.push_back(*it);
+                      *it = -1;
```

Comment:
using -1 as a flag is a bit clumsy.  If you combined index and isVisited you could get rid of it.

Reply:
There is a subtlety here. You need to keep track of two things: 1) the crystals you visit when you perform the clustering algorithm - otherwise you keep looping over the same neighbors forever - and 2) the hits that have been included in a cluster. Note that the isVisited structure needs to be cleared every time the clustering algorithm - a crystal might have two hits at different times and might need to be inspected several times. So the two structures serve different purposes and I don;t think they should be combined.

Reply:
Since you already have a structure that holds the hit indices, the most economical thing to do is reuse it and simply use a special flag to mark a crystal already used.